Change the RDB schema for multi-objective integration

HideakiImamura opened the pull request:
## Motivation
To address #1980, this PR changes the RDB schema and adds the alembic migration code.

## Description of the changes
- Change the RDB schema by fixing `optuna/storages/_rdb/models.py`.
- Fix several unit-tests.
- Fix `optuna/storages/_rdb/alembic/env.py` to autogenerate a migration code including `ALTER TABLE`.
- Add `optuna/storages/_rdb/alembic/versions/v2.4.0.a.py`.

## TODOs
- [x] Test the migration locally.
- [x] Add more unit tests

HideakiImamura commented on optuna/storages/_rdb/alembic/env.py:
```diff
@@ -43,7 +43,9 @@ def run_migrations_offline():
     url = config.get_main_option("sqlalchemy.url")
-    context.configure(url=url, target_metadata=target_metadata, literal_binds=True)
+    context.configure(
+        url=url, target_metadata=target_metadata, literal_binds=True, render_as_batch=True
```
I added `render_as_batch=True` to allow `ALTER TABLE` with sqlite. See [here](https://stackoverflow.com/questions/30378233/sqlite-lack-of-alter-support-alembic-migration-failing-because-of-this-solutio).

toshihikoyanase replied:
[Comment] We may be able to omit `render_as_batch=True` because we use [`Operations.batch_alter_table`](https://alembic.sqlalchemy.org/en/latest/ops.html#alembic.operations.Operations.batch_alter_table). But, I don't think this option causes side-effect and we can keep it as it is.

HideakiImamura commented on optuna/storages/_rdb/alembic/env.py:
```diff
@@ -63,7 +65,9 @@ def run_migrations_online():
     with connectable.connect() as connection:
-        context.configure(connection=connection, target_metadata=target_metadata)
+        context.configure(
+            connection=connection, target_metadata=target_metadata, render_as_batch=True
```
Ditto.

HideakiImamura commented on optuna/storages/_rdb/alembic/versions/v2.4.0.a.py:
```diff
@@ -0,0 +1,80 @@
+
+    with op.batch_alter_table("trial_values", schema=None) as batch_op:
+        batch_op.add_column(sa.Column("objective_id", sa.Integer(), nullable=True))
+        batch_op.create_unique_constraint("value_constraint", ["trial_id", "objective_id"])
```
I manually added the name of the constraint to avoid `ValueError: Constraint must have a name`.
```python
        #before: automatically generated by alembic
        batch_op.create_unique_constraint(None, ["trial_id", "objective_id"])
        
        #after:
        batch_op.create_unique_constraint("value_constraint", ["trial_id", "objective_id"])
```

HideakiImamura commented on optuna/storages/_rdb/storage.py:
```diff
@@ -192,15 +193,24 @@ def set_study_direction(self, study_id: int, direction: StudyDirection) -> None:
         study = models.StudyModel.find_or_raise_by_id(study_id, session)
-
-        if study.direction != StudyDirection.NOT_SET and study.direction != direction:
+        current_direction = tuple(
+            [d.direction for d in models.StudyDirectionModel.where_study(study, session)]
+        )[0]
```
There are a few things I'm tweaking to make it easier to fix later when we do a multi-objective integration, like this.

c-bata commented:
The current migration file only supports schema migration. So the data will be lost.
Could you add data migration?

toshihikoyanase commented on optuna/storages/_rdb/models.py:
```diff
@@ -522,6 +572,58 @@ def all(cls, session: orm.Session) -> List["TrialValueModel"]:
         return session.query(cls).all()
 
 
+class TrialIntermediateValueModel(BaseModel):
```
Do you think if it is possible to add unit tests for `TrialIntermediateValueModel`?

toshihikoyanase commented on optuna/storages/_rdb/storage.py:
```diff
@@ -271,10 +281,12 @@ def get_study_direction(self, study_id: int) -> StudyDirection:
         session = self.scoped_session()
 
         study = models.StudyModel.find_or_raise_by_id(study_id, session)
-        # Terminate transaction explicitly to avoid connection timeout during transaction.
```
I think this comment is for `self._commit(session)` and it still exists. Could you tell me why you remove it?

HideakiImamura replied:
It's my mistake. I will revert the comment.

toshihikoyanase commented on optuna/storages/_rdb/storage.py:
```diff
@@ -192,15 +193,24 @@ def set_study_direction(self, study_id: int, direction: StudyDirection) -> None:
             )
 
-        study.direction = direction
+        session.query(models.StudyDirectionModel).filter(
+            models.StudyDirectionModel.study_id == study_id
+        ).delete()
+
+        direction_model = models.StudyDirectionModel(
+            study_id=study_id, objective_id=0, direction=direction
+        )
+        session.add(direction_model)
```
Why do you remove the study direction entry and re-create the new one?
Can we overwrite the value of `StudyDirectionModel.direction` instead?

toshihikoyanase commented on optuna/storages/_rdb/storage.py:
```diff
@@ -347,23 +359,33 @@ def get_all_study_summaries(self) -> List[StudySummary]:
             best_trial: Optional[models.TrialModel] = None
             try:
-                if study.direction == StudyDirection.MAXIMIZE:
+                if len(directions) > 1:
+                    raise ValueError
```
[Optional] I think these lines will be updated when we integrate multi-objective optimization. It may be kind for readers to add a code comment or an error message about it.

toshihikoyanase reviewed (commented):
Thank you for your PR. I added some early comments. Please take a look.

HideakiImamura commented:
@c-bata @toshihikoyanase I followed all of your reviews. PTAL.

c-bata commented:
Hi, thank you for the update. I have some questions about migration script.

1. Can we ignore the data migration for the existing `optuna.multi_objective` module?
    1. I think it's better to support this because it seems not difficult very much.
1. We've ever implemented reverse migration (e.g. [v1.3.0.a.py](https://github.com/optuna/optuna/blob/aba78bcd14c341c3769bd4657e3c7f7530d7612c/optuna/storages/_rdb/alembic/versions/v1.3.0.a.py#L70-L92)) though it wasn't provided to users. Can we omit the reverse migration from v2.4.0?
    1. I think this is also not difficult, so I recommend you to add reverse migration.

HideakiImamura commented:
Thanks, @c-bata. For your second point, I added the reverse migration code.

For the first point, sorry, I couldn't understand exactly what you meant. The existing multi-objective module employs the `Study` and `FrozenTrial`, so I think we do not need special treatment for existing multi-objective modules. Could you explain a little more?

c-bata commented:
> For the first point, sorry, I couldn't understand exactly what you meant. The existing multi-objective module employs the Study and FrozenTrial, so I think we do not need special treatment for existing multi-objective modules. Could you explain a little more?

Ah, sorry :bow: You are right. I thought that we need to migrate the data in `system_attrs` (like [here](https://github.com/optuna/optuna/blob/efd5e1ac912851030505e383d673530b034fdf7d/optuna/multi_objective/study.py#L131)). But we can't still migrate them now.

HideakiImamura commented:
@c-bata Thanks for the confirmation. So, I think this PR is ready for review. PTAL. Cc: @hvy @toshihikoyanase

toshihikoyanase commented:
@HideakiImamura I created [test scripts](https://gist.github.com/toshihikoyanase/a398d53827fc7597265213b009e43316) and I had an error during migration:

```console
% optuna storage upgrade --storage sqlite:///example.db                                                                                                                                                [~/pfn/code/optuna]
[I 2020-11-25 18:07:24,543] Upgrading the storage schema to the latest version.
[E 2020-11-25 18:07:24,572] (sqlite3.OperationalError) table study_direction already exists
[SQL: 
CREATE TABLE study_direction (
        study_direction_id INTEGER NOT NULL, 
        direction VARCHAR(8) NOT NULL, 
        study_id INTEGER, 
        objective_id INTEGER, 
        PRIMARY KEY (study_direction_id), 
        FOREIGN KEY(study_id) REFERENCES studies (study_id), 
        UNIQUE (study_id, objective_id), 
        CONSTRAINT studydirection CHECK (direction IN ('NOT_SET', 'MINIMIZE', 'MAXIMIZE'))
)

]
(Background on this error at: http://sqlalche.me/e/e3q8)
```
Do you have any ideas about it?

hvy commented on optuna/storages/_rdb/alembic/versions/v2.4.0.a.py:
```diff
@@ -0,0 +1,194 @@
+
+
+# revision identifiers, used by Alembic.
+revision = "v2.4.0.a"
```
We should probably clarify what changes to include together in a release. I believe we at least want the storage interface changes and ideally the entire MO refactoring (up to samplers, excluding pruning) together in order to have a clean release note and to avoid the risk of having to reiterate on the schema between multiple releases. We're currently planning for code freeze Dec. 4 so v2.5 might be safer. Being a bit flexible with v2.4 having it delayed is another option.

HideakiImamura replied:
[Note] In internal discussions, it was decided to delay the release of v2.4 to complete refactoring of the Multi-objective module as much as possible.

hvy commented on optuna/storages/_rdb/models.py:
```diff
@@ -87,6 +86,50 @@ def all(cls, session: orm.Session) -> List["StudyModel"]:
+    study_direction_id = Column(Integer, primary_key=True)
+    direction = Column(Enum(StudyDirection), nullable=False)
+    study_id = Column(Integer, ForeignKey("studies.study_id"))
+    objective_id = Column(Integer)
```
Looks like a primary key at first glance with the id suffix. Do you think that just naming it "objective" is too obfuscated? It could be less confusing.

hvy commented on optuna/storages/_rdb/models.py:
```diff
@@ -522,6 +572,58 @@ def all(cls, session: orm.Session) -> List["TrialValueModel"]:
+    trial_intermediate_values_id = Column(Integer, primary_key=True)
+    trial_id = Column(Integer, ForeignKey("trials.trial_id"))
+    step = Column(Integer)
+    value = Column(Float)
```
How about naming this "intermediate_value"? Seems more consistent (maybe not with the higher layers such as `Trial.report`, but looking at the schema definitions).

hvy replied:
Seems like this comment is unaddressed or did you meant that we shouldn't rename it "intermediate_value"?

HideakiImamura replied:
Ops! I misunderstood that we should `trial_intermediate_values_id` -> `trial_intermediate_value_id`. I will change `value` -> `intermediate_value` now. Thanks for pointing.

HideakiImamura commented:
@toshihikoyanase Thank you for your investigation. I confirmed that the same error you found is observed. I'm not sure now how we resolve it. 🤔

hvy commented on optuna/storages/_rdb/models.py:
```diff
@@ -87,6 +86,50 @@ def all(cls, session: orm.Session) -> List["StudyModel"]:
 
 
+class StudyDirectionModel(BaseModel):
+    __tablename__ = "study_direction"
```
Per convention.
```suggestion
    __tablename__ = "study_directions"
```

toshihikoyanase commented:
Thank you for your investigation. The `RDBStorage` may create a missing table when it loads a study.

https://github.com/optuna/optuna/blob/a59cb935dfb3c2d9a3aba39d15117a407239c9da/optuna/storages/_rdb/storage.py#L114

CC @himkt I think #1957 may be related to this issue. Feel free to add your comments.

toshihikoyanase commented:
@HideakiImamura Thank you for your update. https://github.com/optuna/optuna/pull/2030/commits/9be2587c7b457fed1962eda34715ac114cc44e9b resolved the error. I'll check the change with PostgreSQL and MySQL next.

toshihikoyanase commented:
@HideakiImamura 
I found another error when I upgraded the schema of MySQL. I added the instruction to the [test document](https://gist.github.com/toshihikoyanase/a398d53827fc7597265213b009e43316#mysql), so could you check it?
```console
$ optuna storage upgrade --storage mysql+pymysql://root:test@localhost/optunatest
[I 2020-11-27 11:53:56,550] Upgrading the storage schema to the latest version.
[E 2020-11-27 11:53:56,901] (pymysql.err.IntegrityError) (1062, "Duplicate entry '1' for key 'trial_id'")
[SQL: ALTER TABLE trial_values DROP COLUMN step]
(Background on this error at: http://sqlalche.me/e/gkpj)
```

HideakiImamura commented:
@toshihikoyanase Thanks for the report. I updated the migration code to avoid the reported migration error. I confirmed that https://github.com/optuna/optuna/pull/2030/commits/5fc24bda64c67b80d780f64a7803a61744b7933d passed the test cases you provided. PTAL.

toshihikoyanase commented:
@HideakiImamura Thank you for your swift action. The updated version worked with all three databases. Next, I'll check the downgrade function.

HideakiImamura commented on optuna/storages/_rdb/alembic/versions/v2.4.0.a.py:
```diff
@@ -0,0 +1,124 @@
+
+
+def downgrade():
+    pass
```
I removed the downgrade implementation in this PR since we decide the downgrade method is out of the scope of this PR from the internal discussion. I will work on this in another PR.

hvy replied:
Sounds reasonable given that we'd have to write migration code that mimics the currently trick to allow MO on top of the SO storage interfaces/models.

hvy replied:
A detail but how about leaving a TODO or explain why it's skipped in the code?

HideakiImamura commented:
@toshihikoyanase @hvy I resolved the conflict and remove the downgrade implementation. PTAL.

hvy commented on optuna/storages/_rdb/models.py:
```diff
@@ -87,6 +86,50 @@ def all(cls, session: orm.Session) -> List["StudyModel"]:
+    )
+
+    @classmethod
+    def find_by_study_and_objective(
```
I'm wondering whether we need this method. It's not used (only tested). Did you intend to use it in the near future? I otherwise think we can simplify the API by removing it. Objectives are usually few and `where_study` might do as a generalized version. As a matter of fact, it seems like you're using `where_study` already in the `RDBStorage` followed by indexing `[0]`.

HideakiImamura replied:
Thank you for pointing that out. Surely this API is not necessary. Initially, I added this API as I thought that directions could be given in `Dict[int, StudyDirection]` instead of just sequences and could partially call `set_study_direction` forgiven axes. However, for backward compatibility, we only pass a sequence to `set_study_direction`, so we don't need this API now. I'll delete it.

HideakiImamura replied:
Uh, I've noticed that it is necessary for the MO integration. Please check it in #1994 later. I'll revert the deletion.

HideakiImamura replied:
I committed https://github.com/optuna/optuna/pull/1994/commits/6a564a6a0cdbe71f7cd62f438f0dd35e99676905 and `StudyDirectionModel.find_by_study_and_objective` is used in optuna/storages/_rdb/storage.py:L243-245.

hvy replied:
Sorry for making you go back and forth. Okay, let's leave it as is. Actually, you could write the query out, especially if the method is only used once, but the way we define methods directly in the models aren't clearly communicates so it's not a strong opinion. Let me resolve this one.

hvy commented on optuna/storages/_rdb/storage.py:
```diff
@@ -222,15 +223,22 @@ def set_study_direction(self, study_id: int, direction: StudyDirection) -> None:
             study = models.StudyModel.find_or_raise_by_id(study_id, session)
-
-            if study.direction != StudyDirection.NOT_SET and study.direction != direction:
+            current_direction = tuple(
```
Just curious but is the cast to tuple necessary?

HideakiImamura replied:
When I worked on the MO integration, I assumed that the input was a tuple of `StudyDirection`, so I also cast `current_direction` to a tuple. However, it is not appropriate to make implicit assumptions about the input, so I will cast the input when MO integration. I will simply leave it as a list here.

keisuke-umezawa commented on optuna/storages/_rdb/storage.py:
```diff
@@ -365,6 +378,12 @@ def get_all_study_summaries(self) -> List[StudySummary]:
+                        session.query(models.TrialValueModel)
+                        .filter(models.TrialValueModel.trial_id == best_trial.trial_id)
+                        .filter(models.TrialValueModel.objective == 0)
+                        .one_or_none()
```
[question] What is the criteria to define API in Model? Some query functionalities are implemented as API such as `where_study`, but others are not. For example, these lines are written by the `session.query(...)`.

keisuke-umezawa replied:
A possible criteria is that we define API in Model if we use the implementation more than once.

HideakiImamura replied:
Thank you for the important point. I will not touch `session.query()` directly in `storage.py`, but only in `models.py`. I'm currently touching `session.query()` in `storage.py,` which is inconsistent, so I'll fix it.

HideakiImamura replied:
The part I modified where `session.query()` is used has been rewritten in the method defined in `models.py`. I also rewrote the code parts that could be rewritten with the method in `models.py` without any obvious performance degradation. On the other hand, since it was not obvious how to rewrite it without performance degradation with `get_all_study_summaries` and `get_trials`, I did not rewrite it here. Whether the `session.query()` should only appear in `models.py` is a matter of debate.

hvy replied:
Just to fill in based on what I know, there's currently no real consensus whether queries should be defined via model methods or not. As a matter of fact, we previously discussed getting rid of all these methods to avoid confusion. I think it's reasonable given that many methods are quite short. If there are exceptions we could define functions in `storage.py`, keeping the consistency of the models being free from methods.

HideakiImamura replied:
Thanks for providing the information. It seems that whether we should remove the methods in `models.py` goes beyond the scope of this PR. It's good to discuss it later in another PR or issue.

keisuke-umezawa reviewed (commented):
Add some comments, mainly for Model API.
Did you check that the new DB schema will not degrade any DB performance?

HideakiImamura commented on optuna/storages/_rdb/storage.py:
```diff
@@ -283,8 +291,11 @@ def get_study_direction(self, study_id: int) -> StudyDirection:
             study = models.StudyModel.find_or_raise_by_id(study_id, session)
+            direction = tuple(
+                [d.direction for d in models.StudyDirectionModel.where_study(study, session)]
+            )[0]
```
This is a milestone for the next MO integration. As soon as this PR is merged, we will rename `get_study_direction` as `get_study_directions` and return the output as a tuple, making it multidimensional.

HideakiImamura replied:
I chose `tuple` instead of `list` to be consistent with the existing multi-objective module. In the existing MO module, `MultiObjectiveFrozenTrial.values` is a `tuple` but `MultiObjectiveStudy.directions` is a `list`. After discussing the integration of the MO module with @sile, it's better to unify with a `tuple`. But if `list` is more consistent with the other storage functions, then maybe we should choose `list` for storage. What do you think?

sile replied:
I think either `list` or `tuple` is okay if it's consistent with the other part of the Optuna code base.

HideakiImamura replied:
I will change to use `list` instead of `tuple` for consistency with other storage functions. In PR #1994, to align with the storage functions, how about making the return types of `Study.directions` and `FrozenTrial.values` will be `list`.?

HideakiImamura commented on optuna/storages/_rdb/storage.py:
```diff
@@ -222,15 +223,22 @@ def set_study_direction(self, study_id: int, direction: StudyDirection) -> None:
+                session.query(models.StudyDirectionModel)
+                .filter(models.StudyDirectionModel.study_id == study_id)
+                .first()
+            )
```
We can do it! Great simplification. Thanks.

HideakiImamura commented:
Thanks for the review @keisuke-umezawa! I have addressed all of your reviews. PTAL.

In terms of performance degradation, I have never checked it. I think the performance improvement is important future work, but it does beyond the scope of this PR. I will check the performance and try to improve in other PR. Thanks for the confirmation.

hvy commented on optuna/storages/_rdb/storage.py:
```diff
@@ -584,11 +588,7 @@ def _update_trial(
-                .one_or_none()
-            )
+            trial_model = models.TrialModel.find_or_raise_by_id(trial_id, session)
             if trial_model is None:
                 raise KeyError(models.NOT_FOUND_MSG)
```
Looks like dead code that can be deleted, given your change above to use `find_or_raise_by_id`.
```suggestion
```

hvy commented on optuna/storages/_rdb/storage.py:
```diff
@@ -608,14 +608,19 @@ def _update_trial(
+                        session.add(trial_values_dict[objective])
+                    else:
+                        trial_model.values.extend(
+                            [models.TrialValueModel(objective=objective, value=v)]
```
Oh, the trial_id is automatically added?

HideakiImamura replied:
I think so, but it's good for the code readability to specify the `trial_id`. I will add it.

HideakiImamura replied:
I think so. I implemented this part in reference to the other `Trial***Model`. For example, the extension of the ｀`TrialUserAttributeModel` in the same function. Should I add the `trial_id` for code readability?

hvy replied:
It's a details but yes that'd be clearer. There by the way were other similar occurrences where we explicitly specified the id argument.

toshihikoyanase commented on optuna/storages/_rdb/models.py:
```diff
@@ -512,7 +568,14 @@ def where_study(cls, study: StudyModel, session: orm.Session) -> List["TrialValu
     def where_trial(cls, trial: TrialModel, session: orm.Session) -> List["TrialValueModel"]:
 
-        trial_values = session.query(cls).filter(cls.trial_id == trial.trial_id).all()
+        trial_values = cls.where_trial_id(trial.trial_id, session)
+
+        return trial_values
```
The current `where_trial` is a simple wrapper of `where_trial_id`. This API seems to be redundant, and do we have a plan to revise the model APIs in the future?

HideakiImamura replied:
Yes, I agree with the idea to revise the model APIs. We discussed this point above. Please leave comments if you have or concerns.

toshihikoyanase replied:
Ah, you have already discussed the API design. Let me resolve the comment. Thanks.

hvy commented on optuna/storages/_rdb/storage.py:
```diff
@@ -773,27 +774,12 @@ def _check_and_set_param_distribution(
+                param_name=param_name,
+                param_value=param_value_internal,
+                distribution_json=distributions.distribution_to_json(distribution),
+            ).check_and_add(session)
```
Not introduced by this PR but I'm wondering whether this doesn't cause a violation of the uniqueness constraint on insertion.

hvy commented on optuna/storages/_rdb/storage.py:
```diff
@@ -888,11 +891,7 @@ def get_trial_number_from_id(self, trial_id: int) -> int:
+            trial_model = models.TrialModel.find_or_raise_by_id(trial_id, session)
 
             if not trial_model:
                 raise KeyError("No trial with trial-id {} found.".format(trial_id))
```
Ditto, seems like dead code.

hvy commented on optuna/storages/_rdb/alembic/versions/v2.4.0.a.py:
```diff
@@ -0,0 +1,126 @@
+                sa.Enum("NOT_SET", "MINIMIZE", "MAXIMIZE", name="studydirection"),
+                nullable=False,
+            ),
+            sa.Column("study_id", sa.Integer(), nullable=True),
+            sa.Column("objective", sa.Integer(), nullable=True),
```
Should these fields really be nullable, or is it only for the migration?

HideakiImamura replied:
These fields are automatically generated by alembic. I confirmed that unit tests and migration tests are passed with `nullable = False`. Should we replace `nullable=True` as `nullable=False`?

hvy replied:
I'm wondering if this shouldn't be specified at the model layer.

hvy reviewed (changes requested):
Sorry for the late review, left some additional comments if you could take a look.

HideakiImamura commented on optuna/storages/_rdb/models.py:
```diff
@@ -87,6 +86,49 @@ def all(cls, session: orm.Session) -> List["StudyModel"]:
+    study_direction_id = Column(Integer, primary_key=True)
+    direction = Column(Enum(StudyDirection), nullable=False)
+    study_id = Column(Integer, ForeignKey("studies.study_id"))
+    objective = Column(Integer, default=0)
```
In the case of implementation for single-objective optimization, I set the default to allow you to use this class without specifying an objective. This is used in `storages/_rdb/storage.py:L188` and test functions in `test_models.py`. However, we don't have to set the default value and it's simpler. I will remove the default value.

HideakiImamura commented on optuna/storages/_rdb/storage.py:
```diff
@@ -222,15 +223,17 @@ def set_study_direction(self, study_id: int, direction: StudyDirection) -> None:
-            if study.direction != StudyDirection.NOT_SET and study.direction != direction:
+            current_direction = [
+                d.direction for d in models.StudyDirectionModel.where_study(study, session)
+            ][0]
```
This `[0]` will be removed soon in the continuing MO integration PR #1994, so I preferred to use `where_study` here.

HideakiImamura replied:
Oh, I understood. You mean as follows, right?
```python
n_objectives = len(models.StudyDirectionModel.where_study(study, session)
current_direction = [models.StudyDirectionModel.find_by_study_and_objective(study, i) for i in range(n_objectivers)][0]
```
To compute the number of objectives, we have to use `where_study` anyway. I prefer the current implementation. What do you think?

hvy replied:
You cannot just do `models.StudyDirectionModel.find_by_study_and_objective(study, 0)` not having to construct a temporary list?

HideakiImamura replied:
This is a milestone in #1994 to extend to a multidimensional `StudyDirection`, but it looks good to do so in this PR. I'll fix it.

HideakiImamura commented:
@hvy I have reflected on your reviews. I have commented in replies to the points to be discussed. PTAL.

toshihikoyanase commented on optuna/storages/_rdb/alembic/versions/v2.4.0.a.py:
```diff
@@ -0,0 +1,126 @@
+            sa.Column("trial_intermediate_values_id", sa.Integer(), nullable=False),
+            sa.Column("trial_id", sa.Integer(), nullable=False),
+            sa.Column("step", sa.Integer(), nullable=False),
+            sa.Column("value", sa.Float(), nullable=False),
```
In `TrialIntermediateValueModel`, this column is defined as `intermediate_value`. Could you use the same name to keep consistency?

In addition, `step` and `value` are not nullable, but in `TrialIntermediateValueModel` they are nullable. Could you apply the same constraints to them?

HideakiImamura commented:
@hvy @toshihikoyanase I addressed all of your reviews. PTAL.

hvy commented on optuna/storages/_rdb/alembic/versions/v2.4.0.a.py:
```diff
@@ -0,0 +1,127 @@
+
+    with op.batch_alter_table("trial_values", schema=None) as batch_op:
+        batch_op.add_column(sa.Column("objective", sa.Integer(), nullable=False))
+        batch_op.create_unique_constraint("value_constraint", ["trial_id", "objective"])
```
Memo/TODO: Check name consistency.

hvy commented on optuna/storages/_rdb/alembic/versions/v2.4.0.a.py:
```diff
@@ -0,0 +1,169 @@
+    if "trial_intermediate_values" not in tables:
+        op.create_table(
+            "trial_intermediate_values",
+            sa.Column("trial_intermediate_values_id", sa.Integer(), nullable=False),
```
Ops, seems like a typo.
```suggestion
            sa.Column("trial_intermediate_value_id", sa.Integer(), nullable=False),
```

hvy commented on optuna/storages/_rdb/alembic/versions/v2.4.0.a.py:
```diff
@@ -0,0 +1,169 @@
+                ["trial_id"],
+                ["trials.trial_id"],
+            ),
+            sa.PrimaryKeyConstraint("trial_intermediate_values_id"),
```
Ditto.
```suggestion
            sa.PrimaryKeyConstraint("trial_intermediate_value_id"),
```

hvy commented:
The CI error might be fixed by merging the latest master.

toshihikoyanase reviewed (approved):
I executed the [test scripts](https://gist.github.com/toshihikoyanase/a398d53827fc7597265213b009e43316) again, and I confirmed that the tests passed with SQLite, MySQL, and PostgreSQL. LGTM!

HideakiImamura commented on optuna/storages/_rdb/models.py:
```diff
@@ -33,7 +34,14 @@
 NOT_FOUND_MSG = "Record does not exist."
 
-BaseModel: Any = declarative_base()
+convention = {
+    "ix": "ix_%(column_0_label)s",
+    "uq": "uq_%(table_name)s_%(column_0_name)s",
+    "fk": "fk_%(table_name)s_%(column_0_name)s_%(referred_table_name)s",
+    "pk": "pk_%(table_name)s",
+}
+metadata = MetaData(naming_convention=convention)
+BaseModel: Any = declarative_base(metadata=metadata)
```
See [here](https://alembic.sqlalchemy.org/en/latest/naming.html#integration-of-naming-conventions-into-operations-autogenerate).

HideakiImamura replied:
I did not specify the `ck` since it requires manually add a unique name for each constraint.

HideakiImamura replied:
Introducing this `metadata` may lead to unpredictable migration errors. We decide not to use this now. (I will create another discussion issue.)

hvy reviewed (approved):
LGTM! I think we can merge this PR if @toshihikoyanase 's migration script works with the latest commit.

hvy commented:
Seems to be working with the script, let me merge this PR.
